<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Fabio Morgante &amp; Alex Alekseyenko" />

<meta name="date" content="2024-08-28" />

<title>GWAS tutorial</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">gwas_tutorial</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/fmorgante/gwas_tutorial">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">GWAS tutorial</h1>
<h4 class="author">Fabio Morgante &amp; Alex Alekseyenko</h4>
<h4 class="date">August 28, 2024</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-08-28
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>gwas_tutorial/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240828code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240828)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240828code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240828)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomfmorgantegwastutorialtree94943d942020fd91ce8867b4495620cc742377e8targetblank94943d9a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/fmorgante/gwas_tutorial/tree/94943d942020fd91ce8867b4495620cc742377e8" target="_blank">94943d9</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomfmorgantegwastutorialtree94943d942020fd91ce8867b4495620cc742377e8targetblank94943d9a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/fmorgante/gwas_tutorial/tree/94943d942020fd91ce8867b4495620cc742377e8" target="_blank">94943d9</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    data/.DS_Store
    Ignored:    data/RiceDiversity.44K.germplasm.csv
    Ignored:    data/RiceDiversity_44K_Phenotypes_34traits_PLINK.txt
    Ignored:    data/sativas413.fam
    Ignored:    data/sativas413.map
    Ignored:    data/sativas413.ped

Unstaged changes:
    Modified:   .gitignore
    Modified:   analysis/_site.yml

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/gwas_rice.Rmd</code>) and HTML
(<code>docs/gwas_rice.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/fmorgante/gwas_tutorial/blob/94943d942020fd91ce8867b4495620cc742377e8/analysis/gwas_rice.Rmd" target="_blank">94943d9</a>
</td>
<td>
fmorgante
</td>
<td>
2024-08-28
</td>
<td>
Minor edits
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/fmorgante/gwas_tutorial/d51066270c60de1d28c4a2fde7df6e14275aea22/docs/gwas_rice.html" target="_blank">d510662</a>
</td>
<td>
fmorgante
</td>
<td>
2024-08-28
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/fmorgante/gwas_tutorial/blob/f985e97cdbf0a2cff2cea70ed2b35168fcb10795/analysis/gwas_rice.Rmd" target="_blank">f985e97</a>
</td>
<td>
fmorgante
</td>
<td>
2024-08-28
</td>
<td>
Add rice tutorial
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This tutorial is adapted from <a
href="https://whussain2.github.io/Materials/Teaching/GWAS_R_2.html"
class="uri">https://whussain2.github.io/Materials/Teaching/GWAS_R_2.html</a>.</p>
<div id="description-of-the-data" class="section level2">
<h2>Description of the data</h2>
<p>For this tutorial, we will use rice data that can be dowloaded <a
href="http://ricediversity.org/data/sets/44kgwas/">here</a>. These data
include genotypes at ~44K SNPs and 36 phenotypes for 413 accessions of
<em>Oryza sativa</em>. More details about these data can be found in
this <a href="https://www.nature.com/articles/ncomms1467">paper</a>.</p>
</div>
<div id="load-libraies-needed" class="section level2">
<h2>Load libraies needed</h2>
<p>First, we will load the necessary R packages, BGLR, SNPRelate, dplyr,
rrBLUP, and qqman.</p>
<pre class="r"><code>library(BGLR)
library(SNPRelate)
library(dplyr)
library(rrBLUP)
library(qqman)</code></pre>
</div>
<div id="prepare-genotype-and-phenotype-data" class="section level2">
<h2>Prepare genotype and phenotype data</h2>
<p>The genotype data are included in three files with extension .fam
(which includes information about the samples), .map (which contains
information about the SNPs, such as chromosome and position), .ped
(which contains the genotype calls, where 0 and 3 mean homozygous for
either allele, 1 means heterozygous, and 2 means missing). For our
purpose, we will recode missing geneotype as NA and heterozygous
genotypes as 1.</p>
<pre class="r"><code>###Load genotype file
geno &lt;- read_ped(&quot;data/sativas413.ped&quot;)
FAM &lt;- read.table(&quot;data/sativas413.fam&quot;, header=FALSE, sep=&quot; &quot;)
MAP &lt;- read.table(&quot;data/sativas413.map&quot;, header=FALSE, sep=&quot;\t&quot;)

p &lt;- geno$p
n &lt;- geno$n

###Access genotype matrix
geno_vec &lt;- geno$x

###Recode genotypes
geno_vec[geno_vec == 2] &lt;- NA  # Converting missing data to NA
geno_vec[geno_vec == 3] &lt;- 2  # Converting 3 to 2

###Convert the genotype data into matrix, transpose and check dimensions
geno_mat &lt;- matrix(geno_vec, nrow = p, ncol = n, byrow = TRUE)
geno_mat &lt;- t(geno_mat)
dim(geno_mat)</code></pre>
<pre><code>[1]   413 36901</code></pre>
<pre class="r"><code>colnames(geno_mat) &lt;- MAP$V2
rownames(geno_mat) &lt;- FAM$V2

dim(geno_mat)</code></pre>
<pre><code>[1]   413 36901</code></pre>
<pre class="r"><code>geno_mat[1:5, 1:5]</code></pre>
<pre><code>  id1000001 id1000003 id1000005 id1000007 id1000008
1         0         0         0         0         0
3         2         2         0         2         2
4         2         2         0         2         2
5         2         2         2         0         2
6         2         2         0         2         2</code></pre>
<p>From the phenotype data, we selected Flowering.time.at.Aberdeen as
our trait of interest. Note that NSFTVID is the ID of the accession and
missing value are coded as NA. Accessions with missing phenotype are
removed.</p>
<pre class="r"><code>###Load phenotype data
rice_pheno &lt;- read.table(&quot;data/RiceDiversity_44K_Phenotypes_34traits_PLINK.txt&quot;, 
                         header = TRUE, stringsAsFactors = FALSE, sep = &quot;\t&quot;)

###Extract phenotype and remove missing values
y &lt;- rice_pheno$Flowering.time.at.Aberdeen
index &lt;- !is.na(y)
y &lt;- data.frame(NSFTV.ID = rice_pheno$NSFTVID[index], y = y[index])

dim(y)</code></pre>
<pre><code>[1] 359   2</code></pre>
<pre class="r"><code>head(y)</code></pre>
<pre><code>  NSFTV.ID   y
1        1  81
2        3  83
3        4  93
4        5 108
5        6 101
6        7 158</code></pre>
<p>Let’s now do some QC on the genotype data. This includes removing
accessions with missing phenotype, computing minor allele frequency
(MAF) for each SNP, and removing SNPs with MAF smaller than 0.05.</p>
<pre class="r"><code>###Remove accessions with missing pheno
geno_mat &lt;- geno_mat[index, ]

###Filter out variants with small MAF
af &lt;- colMeans(geno_mat, na.rm = TRUE)/2
maf &lt;- pmin(af, 1-af)
to_remove &lt;- which(maf &lt; 0.05)
geno_mat &lt;- geno_mat[, -to_remove]
MAP &lt;- MAP[-to_remove, ]

dim(geno_mat)</code></pre>
<pre><code>[1]   359 33572</code></pre>
</div>
<div id="investigation-of-population-structure" class="section level2">
<h2>Investigation of population structure</h2>
<p>These accessions come from different populations. The presence of
population structure can induce false positives, so we need to
investigate this. To do so, we will build a genomic relationship matrix
(GRM) that measures the similarity between accessions based genotypes.
The higher the coefficient of genomic relationship, the more similar two
accessions are.</p>
<pre class="r"><code>###Create gds format file and save it as 44k.gds
snpgdsCreateGeno(&quot;data/44k.gds&quot;, genmat = geno_mat, sample.id = rownames(geno_mat), snp.id = colnames(geno_mat), 
                 snp.chromosome = MAP$V1, snp.position = MAP$V4, snpfirstdim = FALSE)
# Now open the 44k.gds file
geno_44k &lt;- snpgdsOpen(&quot;data/44k.gds&quot;)
snpgdsSummary(&quot;data/44k.gds&quot;)</code></pre>
<pre><code>The file name: /Users/fabiom/Documents/GIT/gwas_tutorial/data/44k.gds 
The total number of samples: 359 
The total number of SNPs: 33572 
SNP genotypes are stored in SNP-major mode (Sample X SNP).</code></pre>
<pre class="r"><code>###Compute GRM
grm_obj &lt;- snpgdsGRM(geno_44k, method=&quot;GCTA&quot;)</code></pre>
<pre><code>Genetic Relationship Matrix (GRM, GCTA):
Excluding 0 SNP on non-autosomes
Excluding 0 SNP (monomorphic: TRUE, MAF: NaN, missing rate: NaN)
    # of samples: 359
    # of SNPs: 33,572
    using 1 thread
GRM Calculation:    the sum of all selected genotypes (0,1,2) = 7149883
CPU capabilities:
Wed Aug 28 10:49:32 2024    (internal increment: 1368)
[..................................................]  0%, ETC: ---        [==================================================] 100%, completed, 2s
Wed Aug 28 10:49:34 2024    Done.</code></pre>
<pre class="r"><code>GRM &lt;- grm_obj$grm
colnames(GRM) &lt;- rownames(GRM) &lt;- grm_obj$sample.id

dim(GRM)</code></pre>
<pre><code>[1] 359 359</code></pre>
<pre class="r"><code>GRM[1:5, 1:5]</code></pre>
<pre><code>           1          3          4          5          6
1  1.3019512 -1.0096800 -0.8575139  0.0078842 -0.8105344
3 -1.0096800  3.1219327  0.7288741 -0.2955218  0.6856645
4 -0.8575139  0.7288741  2.8092981 -0.1398424  2.3414236
5  0.0078842 -0.2955218 -0.1398424  1.7246536 -0.1352005
6 -0.8105344  0.6856645  2.3414236 -0.1352005  2.7222408</code></pre>
<p>We will perform principal component analysis (PCA) via eigen
decomposition of the GRM, to determine whether population structure is
actually present.</p>
<pre class="r"><code>###Perform eigen decomposition
eig &lt;- eigen(GRM)

eig_vectors &lt;- eig$vectors
colnames(eig_vectors) &lt;- paste0(&quot;EV&quot;, 1:ncol(eig_vectors))
eig_vectors_df &lt;- data.frame(NSFTV.ID = grm_obj$sample.id, eig_vectors)

###Plot PC1 vs PC2 and label points according to the pop they belong to
plot(eig_vectors_df$EV1, eig_vectors_df$EV2, xlab = &quot;PC1&quot;, ylab = &quot;PC2&quot;)</code></pre>
<p><img src="figure/gwas_rice.Rmd/GRM%20PCA%201-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-GRM-PCA-1-1">
Past versions of "GRM PCA 1-1.png"
</button>
</p>
<div id="fig-no-spaces-GRM-PCA-1-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/fmorgante/gwas_tutorial/blob/d51066270c60de1d28c4a2fde7df6e14275aea22/docs/figure/gwas_rice.Rmd/GRM PCA 1-1.png" target="_blank">d510662</a>
</td>
<td>
fmorgante
</td>
<td>
2024-08-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>From the plot of PC1 vs PC2, we can see that there is indeed
population structure, with groups of accessions being closer to each
other than to the rest. This becomes even more clear when we overlay
information about the population that each accession comes from.</p>
<pre class="r"><code>###Add population info from the web to PC data
pop_info &lt;- read.csv(&quot;data/RiceDiversity.44K.germplasm.csv&quot;, 
                     header = TRUE, skip = 1, stringsAsFactors = FALSE)
pop_info &lt;- pop_info[, c(&quot;NSFTV.ID&quot;, &quot;Sub.population&quot;)]
pop_info$NSFTV.ID &lt;- as.character(pop_info$NSFTV.ID)

eig_vectors_info &lt;- left_join(eig_vectors_df, pop_info, by=&quot;NSFTV.ID&quot;)

###Plot PC1 vs PC2 and label points according to the pop they belong to
plot(eig_vectors_info$EV1, eig_vectors_info$EV2, xlab = &quot;PC1&quot;, ylab = &quot;PC2&quot;, col = c(1:6)[factor(eig_vectors_info$Sub.population)])
legend(x = &quot;topright&quot;, legend = levels(factor(eig_vectors_info$Sub.population)), col = c(1:6), 
       pch = 1, cex = 0.6)</code></pre>
<p><img src="figure/gwas_rice.Rmd/GRM%20PCA%202-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-GRM-PCA-2-1">
Past versions of "GRM PCA 2-1.png"
</button>
</p>
<div id="fig-no-spaces-GRM-PCA-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/fmorgante/gwas_tutorial/blob/d51066270c60de1d28c4a2fde7df6e14275aea22/docs/figure/gwas_rice.Rmd/GRM PCA 2-1.png" target="_blank">d510662</a>
</td>
<td>
fmorgante
</td>
<td>
2024-08-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="genome-wide-association-study" class="section level2">
<h2>Genome-Wide Association Study</h2>
<p>We will now perform a GWAS for our trait of interest using a linear
mixed model that models the covariance between accessions (our random
effect) using the GRM. That way, we correct for population
structure.</p>
<pre class="r"><code>###Make the GRM positive semi-definite
eig_values &lt;- eig$values
eig_values[eig_values&lt;0] &lt;- 0
GRM_SPD &lt;- eig_vectors %*% diag(eig_values) %*% t(eig_vectors)

###Prepare genotype data in the format expected by rrBLUP
geno_final &lt;- data.frame(marker = MAP[, 2], chrom = MAP[, 1], pos = MAP[, 4], 
                         t(geno_mat - 1), check.names = FALSE)  # W = \in{-1, 0, 1}

###Perform GWAS via linear mixed model
results &lt;- GWAS(pheno=y, geno=geno_final, K=GRM_SPD, min.MAF=0, P3D = TRUE, plot = FALSE)</code></pre>
<pre><code>[1] &quot;GWAS for trait: y&quot;
[1] &quot;Variance components estimated. Testing markers.&quot;</code></pre>
<pre class="r"><code>results$p_value &lt;- 10^(-results$y)</code></pre>
<p>The results are displayed as a Q-Q plot, where the observed p-values
are plotted against the expected p-values under the null hypothesis of
no association between the SNP and the phenotype.</p>
<pre class="r"><code>###Make qqplot
qq(results$p_value)</code></pre>
<p><img src="figure/gwas_rice.Rmd/QQ%20LMM-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-QQ-LMM-1">
Past versions of "QQ LMM-1.png"
</button>
</p>
<div id="fig-no-spaces-QQ-LMM-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/fmorgante/gwas_tutorial/blob/d51066270c60de1d28c4a2fde7df6e14275aea22/docs/figure/gwas_rice.Rmd/QQ LMM-1.png" target="_blank">d510662</a>
</td>
<td>
fmorgante
</td>
<td>
2024-08-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This plot looks pretty good – large observed p-values are well
aligned to their expectation under the null, while smaller p-valuesshow
a deviation from their expectation under the null, hinting at
discovering true signals. We then display the final results as a
manhattan plot, where the significance line reflects a Bonferroni
correction.</p>
<pre class="r"><code>###Make qqplot
manhattan(results, chr=&quot;chrom&quot;, bp=&quot;pos&quot;, p=&quot;p_value&quot;, snp = &quot;marker&quot;, 
          genomewideline = -log10(0.05/nrow(geno_final)), suggestiveline=FALSE)</code></pre>
<p><img src="figure/gwas_rice.Rmd/man%20LMM-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-man-LMM-1">
Past versions of "man LMM-1.png"
</button>
</p>
<div id="fig-no-spaces-man-LMM-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/fmorgante/gwas_tutorial/blob/d51066270c60de1d28c4a2fde7df6e14275aea22/docs/figure/gwas_rice.Rmd/man LMM-1.png" target="_blank">d510662</a>
</td>
<td>
fmorgante
</td>
<td>
2024-08-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>It looks like there is a strong association signal on chromosome 6,
and a weaker one on chromosome 1.</p>
<p>As an exercise, what if we had done GWAS with a regular linear model
disregarding population structure?</p>
<pre class="r"><code>###Perform GWAS via simple linear regression
results$p_value_lm &lt;- as.numeric(NA)

# This is slow code that is only useful for the teaching purpose 
for(i in 1:nrow(geno_final)){
  fit &lt;- lm(y$y ~ as.numeric(geno_final[i, 4:ncol(geno_final)]))
  results[i, &quot;p_value_lm&quot;] &lt;- summary(fit)$coefficients[2,4]
}

qq(results$p_value_lm)</code></pre>
<p><img src="figure/gwas_rice.Rmd/GWAS%20LM-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-GWAS-LM-1">
Past versions of "GWAS LM-1.png"
</button>
</p>
<div id="fig-no-spaces-GWAS-LM-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/fmorgante/gwas_tutorial/blob/d51066270c60de1d28c4a2fde7df6e14275aea22/docs/figure/gwas_rice.Rmd/GWAS LM-1.png" target="_blank">d510662</a>
</td>
<td>
fmorgante
</td>
<td>
2024-08-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The Q-Q plot looks pretty bad now, with clear signs of uncorrected
confounding.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.1 (2022-06-23)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Monterey 12.7.6

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] qqman_0.1.9      rrBLUP_4.6.3     dplyr_1.1.4      SNPRelate_1.32.2
[5] gdsfmt_1.34.1    BGLR_1.1.2       workflowr_1.7.1 

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.12       highr_0.11        compiler_4.2.1    pillar_1.9.0     
 [5] bslib_0.8.0       later_1.3.2       git2r_0.33.0      jquerylib_0.1.4  
 [9] tools_4.2.1       getPass_0.2-4     digest_0.6.35     jsonlite_1.8.8   
[13] evaluate_0.24.0   lifecycle_1.0.4   tibble_3.2.1      pkgconfig_2.0.3  
[17] rlang_1.1.3       cli_3.6.2         rstudioapi_0.16.0 parallel_4.2.1   
[21] yaml_2.3.8        xfun_0.44         fastmap_1.2.0     httr_1.4.7       
[25] stringr_1.5.1     knitr_1.48        generics_0.1.3    fs_1.6.4         
[29] vctrs_0.6.5       sass_0.4.9        tidyselect_1.2.1  rprojroot_2.0.4  
[33] calibrate_1.7.7   glue_1.7.0        R6_2.5.1          processx_3.8.4   
[37] fansi_1.0.6       rmarkdown_2.28    callr_3.7.6       magrittr_2.0.3   
[41] whisker_0.4.1     MASS_7.3-57       ps_1.7.6          promises_1.3.0   
[45] htmltools_0.5.8.1 httpuv_1.6.15     utf8_1.2.4        stringi_1.8.4    
[49] truncnorm_1.0-9   cachem_1.1.0     </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
